src/Expander.hs

```diff
@@ -1072,26 +1074,25 @@ addRootScope' stx = do
   return (addScope' stx rsc)
 
 
-forkExpandOneDecl :: DeclPtr -> Scope -> DeclValidityPtr -> Syntax -> Expand ()
-forkExpandOneDecl dest sc phaseDest stx = do
-  forkExpanderTask $ ExpandDecl dest sc stx phaseDest
+forkExpandOneDecl :: ModBodyPtr -> Scope -> DeclValidityPtr -> Syntax -> Expand ()
+forkExpandOneDecl dest sc declValidityPtr stx = do
+  forkExpanderTask $ ExpandModBody dest sc stx declValidityPtr
```

Comment:
I want to change the name of `forkExpandOneDecl` and `forkExpandDecls`, but I'm not sure to what.

Reply:
How about "forkExpandDecl" and "forkExpandDeclGroup"?

Reply:
`forkExpandOneDecl` is no longer expanding a single declaration, so I don't think either `forkExpandOneDecl` nor `forkExpandDecl` is appropriate. Both functions are expanding a DeclTree... but one of the two takes an extra `DeclValidityPtr` argument. I am hoping that the difference between those two functions will become clearer to me once I make sure all the names in the left branch are bound before expanding the right branch.

Reply:
As I was hoping, that work helped me figure it out: in addition to declarations and declaration trees, there is a third concept called a declaration _form_. `(define ...)`, `(no-decls)`, and `(decls ...)` are all declaration forms, not declaration trees, and each of them expands to a tree declaration, which contains zero or more declarations. Unlike declaration trees, which are can be grouped into a tree structure, declaration forms can only be grouped into _lists_. Some positions expect a single declaration form, while other positions expect a list of zero or more declaration forms.

In addition to that, there are also _dependent_ declaration forms. A dependent declaration form is just like a regular declaration form, except its corresponding task must wait until some other `DeclValidityPtr` gets filled before the task can begin.

Before this PR, there was a coincidence between the two distinctions: the position at the head of a list always held a single declaration and was always non-dependent, while the position at the tail of a list always held a list of declarations and was always dependent on the head. With declaration trees, however, the coincidence no longer holds: the two positions inside a `decls`  both hold a single declaration, while only the second depends on the first. The tail of a list of declarations is still an example of position which holds a list of declarations and is dependent on the head, while the position inside a `meta` holds a list a declaration but is not dependent.

So I think `forkExpandOneDecl` and `forkExpandDecls` need to become _four_ different functions:

1. `forkExpandDeclForm`
2. `forkExpandDeclForms`
3. `forkExpandDependentDeclForm`
4. `forkExpandDependentDeclForms`

Furthermore, `(decls (no-decls) (define ...))` demonstrates that it is now possible for a dependent declaration form to depend on a declaration form which ends up binding zero identifiers, I think there was another coincidence between the part of `DeclValidityPtr` which indicates that a dependent task can proceed, and the part which indicates which phase each identifier is valid in. I'm not sure what to call the two parts yet.

Reply:
Hmm, another difference between `forkExpandOneDecl` and `forkExpandDecls` is that the former takes a `Scope` argument. I don't understand why the places which only accept a single declaration form inherit a `Scope` from elsewhere, while the places which accept multiple declaration forms create on `Scope` for each entry in the list. I would have expected each binding site to generate a new `Scope`, but instead it looks like all the primitives in `declPrims` receive that new `Scope` as an argument. I think that's because of yet another coincidence: each entry in the list of declaration form binds exactly one name, and so there is no need to wait to see if the head declaration form is a `define` or something else, `forkExpandDecls` already knows that it will bind a name and so it pre-emptively generates a `Scope`. This way, it can hand that `Scope` to the tail.

I guess that answers the question of how to name the part of `DeclValidityPtr` which indicates that a dependent task can proceed: a `ScopePtr`!

Reply:
What I like less about this is that it's not a `Scope` created as a result of user macro expansion, and this name doesn't communicate that it exists mainly for synchronization purposes.

Reply:
> this name doesn't communicate that it exists mainly for synchronization purposes.

Hmm, don't all our `Ptr`s exist for synchronization purposes? Each `Ptr` is a place where we'd normally do a call, but instead we schedule a task, and instead of doing something with the result, we schedule another task which is blocked until that result is available.

In this case, I previously thought that these DeclValidityPtrs were a special case, because the code which waits for that Ptr doesn't necessarily need to know the value it points to, it simply waits on it in order to make sure that some global map of names has been updated with all the names it might use.

But now that I have realized that we need to thread a ScopeSet (I said Scope before, but there could be multiple declarations each introducing their own Scope, so it's a ScopeSet), the picture is much more uniform: the code which waits for that ScopeSetPtr is waiting for it because it needs the ScopeSet it points to. That ScopeSet is conceptually an environment mapping all the names which are in scope to (a pointer to) their values, except we've implemented that via a more complicated setup whereby you can get access to that environment from the global environment by tagging your name lookups with a ScopeSet. So obviously we shouldn't fill a ScopeSetPtr until after we've updated that global environment.

In conclusion, I think `ScopeSetPtr` is a pretty descriptive name to indicate that we're waiting until all the names in the corresponding environment are bound!

Reply:
I have now split DeclValidityPtr into PhaseValidityPtr and ScopeSetPtr, but I don't think I'll push my changes, as I no longer think they need to be split.

My understanding is that before this PR, as we traverse a list of declarations, we allocate one Scope per declaration and then we add that Scope to the Syntax of the remaining declarations at the right phase (or possibly all phases in the case of `import`). These two pieces of information, the Scope and the phase, are transmitted via two very different paths. The Scope is created first, in `forkExpandDecls`, and is passed as a parameter of both the `ExpandDecl` and the `ExpandMoreDecls` tasks. This allows the `ExpandDecl` task to bind its identifier to the same Scope which `ExpandMoreDecls` will add to the Syntax of the remaining declarations. For the phase, `forkExpandDecls` instead creates a new `DeclValidityPtr` which the `ExpandDecl` task fills and which the `ExpandMoreDecls` task waits upon.

A different design would be to transfer both the Scope and the phase via the same path: `forkExpandDecls` would only create a new `DeclValidityPtr`, not a new Scope, and then the `ExpandDecl` task would fill that `DeclValidityPtr` with _both_ a Scope and a phase.

That design generalizes better to the declaration-tree case, because a declaration form can now bind an arbitrary number of identifiers and thus needs to be able to allocate an arbitrary number of Scopes, so the `ExpandDecl` cannot simply receive a single Scope as input. Instead, it must now be able to fill the `DeclValidityPtr` with zero or more Scopes, each for each of them, it must specify whether the scope is valid in a particular phase or in all phases. Put another way, the thing which `DeclValidityPtr` points to is map from AllPhases to a set of Scopes, and a map from each Phase to a set of Scopes. That is, it points to a `ScopeSet`!

Now that I no longer want to split `DeclValidityPtr` into two parts, I don't really need to rename it, but I still think `ScopeSetPtr` would be a better name than `DeclValidityPtr` if it points to a ScopeSet.

Reply:
I now realize that I don't really understand the "it's not a Scope created as a result of user macro expansion" part of your objection to the name, can you explain why the name ScopePtr implies a relation with user macros?

Reply:
Only that when we have used `Ptr`s elsewhere, it's been to make an incrementally-built structure that comes out of the expander. But that's not really inherent to pointerness. I retract my objection :-)